feat: 프로필 페이지 API 연동

yeni023 opened the pull request:
## 🔗 반영 브랜치
- closed #119
- `yeeun/Profile-api → dev`

## 📝 작업 내용
프로필 페이지의 API 연동 작업을 진행했습니다.
- 닉네임 조회 및 수정 기능 연동
- 프로필 이미지 조회, 업로드, 삭제 기능 연동
- 삭제 후에도 수정(Edit) 가능한 상태 유지
- 삭제 및 수정 시 `/users/me` API를 refetch하여 최신 상태 반영

### 📸 스크린샷 (선택)
<img width="886" alt="스크린샷 2025-05-14 오후 2 26 18" src="https://github.com/user-attachments/assets/360a6144-ad14-40e6-bafe-f66bf0d47693" />

## 💬 리뷰 요구사항(선택)
- 닉네임 수정 시 한 번 새로고침을 해야 반영되는 문제가 있어 추후 수정 예정입니다.
- 연동한 API 동작은 테스트 시 문제 없었지만, 한 번 더 확인 부탁드립니다.
- `mutations`, `queries` 사용이 아직 익숙하지 않아 비효율적인 부분이 있을 수 있습니다. 피드백 주시면 감사하겠습니다.

yeni023 committed:
feat: 프로필 페이지 API 연동 및 이미지 수정 기능 개선

- 기본 프로필 이미지를 프론트 고정 경로 대신 API 응답 기반으로 렌더링
- 이미지 업로드 시 avatarUrl을 백엔드 응답값으로 갱신
- 이미지 삭제 후 /users/me API refetch로 최신 상태 반영
- 삭제 이후에도 이미지 수정(Edit)이 가능하도록 input 초기화
- useUsersMe를 활용하여 프로필 정보 최신화 처리
- 로그아웃 시 router 전환 대신 window.location.href로 이동 처리

yeni023 committed:
style: 프로필 수정 버튼 hover 시 커서 pointer 적용

HJ0311 commented on src/pages/ProfilePage/components/logout-confirm-modal/index.tsx:
```diff
@@ -8,12 +11,29 @@ type Props = {
+      },
+      onError: (error) => {
+        console.error('로그아웃 실패:', error);
+        alert('로그아웃에 실패했습니다.');
```
이 부분 영어로 통일하면 좋을 것 같아요!

HJ0311 commented on src/pages/ProfilePage/index.tsx:
```diff
@@ -9,19 +12,34 @@ import ProfileImage from './components/profile-image';
+        navigate('/signin');
+      },
+      onError: () => {
+        alert('로그아웃에 실패했습니다.');
```
이 부분도 영어로 통일하면 좋을 것 같아요

HJ0311 commented on src/pages/ProfilePage/index.tsx:
```diff
@@ -9,19 +12,34 @@ import ProfileImage from './components/profile-image';
   };
 
+  if (isLoading) return <div className={styles.container}>Loading...</div>;
+  if (isError || !data?.data) return <div className={styles.container}>불러오기 실패</div>;
```
이 부분도 영어로 통일하면 좋을 것 같아요

HJ0311 reviewed (dismissed):
수고하셨습니다!

hakilee commented on src/api/fetcher.ts:
```diff
@@ -14,7 +14,8 @@ const defaultOption: Options = {
 export const instance = ky.create({
   prefixUrl: '/api',
-  headers: { 'content-type': 'application/json' },
+  // application/json을 고정으로 지정하고 있어서, FormData를 보낼 때 문제 발생하여 아래 코드 주석 처리
```
이건 고정으로 가져가는 것이 맞는 것 같습니다. FormData를 보내는 Mutation에서 `content-type`을 overriding 해주세요.

hakilee commented on src/api/mutations/index.ts:
```diff
@@ -1,3 +1,6 @@
 export * from './useLogin';
 export * from './useMailVerification';
 export * from './useRegister';
+// ProfilePage
```
주석 없이 문단을 구별하는 것만으로도 도메인이 구분되지 않을까요?! 필요하다고 느끼시면 4번째 줄을 더 띄워야 할 것 같아요.

```suggestion

// ProfilePage
```

yeni023 replied:
문단으로도 충분할 것 같습니다! 주석 지우고 올릴게요!

hakilee commented on src/api/mutations/useLogout.ts:
```diff
@@ -0,0 +1,9 @@
+import { fetcher } from '@/api';
+
+const postLogout = async (): Promise<void> => {
+  await fetcher.post<void>('auth/logout', { json: {} });
```
```suggestion
  await fetcher.post<void>('auth/logout');
```

`json: {}` 부분이 꼭 필요하지 않다면 지워주세요.

hakilee commented on src/api/mutations/useUpdateProfile.ts:
```diff
@@ -0,0 +1,13 @@
+
+export const useUpdateProfile = () =>
+  useMutation({
+    mutationFn: (body: Request) => fetcher.patch<HttpResponse<User>>('users/me', { json: body }),
```
`patchUpdateProfile` 과 같은 이름의 함수로 분리해서 다른 뮤테이션 파일과 일관성을 맞춰주세요.

hakilee commented on src/api/mutations/useUploadAvatar.ts:
```diff
@@ -0,0 +1,14 @@
+};
+
+const postAvatar = (formData: FormData) =>
+  fetcher.post<{ data: Response }>('users/avatar', {
```
이 부분에 headers를 줄 수 있는 옵션이 있어요. `@/api/fetcher.ts` 파일의 하단을 보면 `post: <T>(pathname: string, options?: Options) => resultify<T>(instance.post(pathname, options))`로 구성되어 있습니다.

여기에서 options을 수정해서 `content-type`를 변경해주세요.

hakilee commented on src/pages/ProfilePage/components/edit-nickname-modal/index.tsx:
```diff
@@ -1,6 +1,8 @@
 import { useState } from 'react';
 import ReactDOM from 'react-dom';
 
+import { useUpdateProfile } from '@/api/mutations/useUpdateProfile';
```
`@/api/mutations/index.ts` 파일에 `export`을 추가한다면 from 뒤가 더 짧아질 것 같습니다.

hakilee commented on src/pages/ProfilePage/components/profile-image/index.tsx:
```diff
@@ -1,52 +1,75 @@
   };
 
-  const handleFileChange = async (event: React.ChangeEvent<HTMLInputElement>) => {
+  const handleFileChange = (event: React.ChangeEvent<HTMLInputElement>) => {
```
```suggestion
  const handleFileChange = (event: ChangeEvent<HTMLInputElement>) => {
```

hakilee reviewed (commented):
고생하셨습니다~ 몇 가지 통일을 하면 좋을 것 같은데요,

Mutation을 사용할 때, `const { mutate: name } = use~Mutation` 문법을 사용하게 되는데, 이때 name의 끝에는 Mutation을 붙여주세요. 예를 들어 로그인 작업이라면,

```typescript
const { mutate: useLoginMuatation } = useLogin();
```
이 되는 거죠.

파일과 코드가 많아질수록 이러한 룰을 설정해야 어떤 훅인지, 어떤 작업이 이루어지는지를 명확하게 알 수 있을 것 같아요.

현재 모달이나 일부 페이지가 `vanilla-extract`가 아닌, `emotion`이 사용되고 있어요. 이번 작업은 API 연동 작업이니, 디자인을 수정하는 작업이 추가로 필요할 듯싶네요!

마지막으로, `content-type`은 제거하지 않는 것이 좋겠습니다. 예외 사항이 필요한 경우에만 헤더를 오버라이딩 해서 다른 값을 주는 것이 더 좋을 것 같고, 이를 위해서 옵션을 추가로 받으니까요.

yeni023 committed:
refactor: 뮤테이션 네이밍 통일 및 fetcher 내부 FormData 처리 로직 반영

- useMutation 구조 분해 시 네이밍을 ~Mutation 형태로 통일
- useUpdateProfile 등 익명 함수를 명시적인 함수로 분리
- '@/api/mutations'에서 export 정리하여 import 경로 단순화
- FormData 사용 시 content-type 제거 로직을 fetcher 내부에서 자동 처리되도록 수정
- 불필요한 빈 JSON body 제거 (logout)

hakilee committed:
refactor: improve profile page and api mutations

this commit refactors the profile page and related api mutations to improve user experience and code maintainability.

- removes unnecessary navigate from profile page and uses window.location.href for logout redirection.
- handles null values for win, lose, and tournament stats in the profile page.
- simplifies post method in fetcher.ts.
- modifies mutation functions (login, logout, mailverification, register, updateprofile, uploadavatar, deleteavatar) to use consistent payload naming and improve query invalidation after mutations.
- removes unnecessary comments in user.ts.

hakilee committed:
refactor: remove unnecessary refetch after avatar mutation

this change removes the explicit `refetch` calls after successful avatar upload and deletion. the user data is now fetched and updated automatically. this simplifies the code and prevents potential race conditions.

hakilee reviewed (approved):
API 관련해서 일부 파일 수정했습니다. 모달 관련해서 모두 vanilla-extract 스타일링을 적용하면 될 것 같아요! 고생하셨습니다.
